### DISCUSSION TITLE
Scaling background image

### DISCUSSION THREAD

**Discussion by pgzh** [jcorporation/myMPD#506] [Ideas]
I'd like to suggest a possible enhancement:
If a background image is selected it would be nice if it were scaled to fill the background area completely.

Depending on the width/height ratio, there may be an area above and below or left and right where the bgimage isn't displayed.

If the height of the image is smaller than the display area (or more precisely the width/height ratio of the image is greater than that of the display), maybe it could be scaled to full height and cropped left and right.
If the width is smaller than the display area (width/height ratio smaller than that of the display), it could be scaled to full width and cropped top and bottom.

Especially on mobile devices with rotating displays the whole area would be covered by the bgimage this way.

On top of that it would be really cool if the bgimage would extend to the whole display and the bars at the top and the bottom could be configured with adjustable transparency.

Thanks for making this great mpd client!!

**Comment by jcorporation** [comment]
The css for the background sizing is:
```
background-position: center center; 
background-size: contain;  
background-repeat: no-repeat;
```

Transparency for the toolbars could be an option.

**Comment by pgzh** [comment]
Hi, thank you for the quick response.
Looking at the css I think
`background-size: cover;`
may just achieve what I tried to describe.
contain will ensure the whole image is displayed and may add bars, but cover will keep aspect ratio, scale to use all space and avoid bars but instead crop part of the image instead.

So if you'd consider transparency maybe you could also consider a config option to choose between contain and cover for the background-size property?

**Reply by jcorporation** [reply]
This should be no problem.

**Comment by pgzh** [comment]
I'd love to see this implemented, thanks for being open to my suggestions!
And one more question: can the css filter applied to the cover art also be applied to the bgimage?

**Reply by jcorporation** [reply]
Css filters for background images (backdrop-filter) are not supported by all major browsers.

**Reply by pgzh** [reply]
Too bad...how is the cover art 'background' different then? I imagined the bgimage was just a static version of the cover art background actually.

**Reply by jcorporation** [reply]
Albumart background is implemented differently, because of the fading effect. 
The static image is an background image, the albumart are background div elements.

**Reply by pgzh** [reply]
So wouldn't it be possible to use a static image for the albumart background div elements to create a more flexible background image?
Nobody will probably want to enable both at the same time, so the background feature could just set the image used for the bg div elements and the css feature could be used as well.

**Reply by jcorporation** [reply]
It is possible, but I see no benefit. The static background image can be edited with image editing software before.

**Reply by pgzh** [reply]
Well, you've got a point there - absolutely right.
But there is another observation I made regarding this:
On my mobile phone when scrolling through the queue, the background image is visible only behind a few tracks in the middle of the page (about 10 of 100). So it's like it has a fixed position in the page and just sits there at one position in the tracklist.

The cover art background however is always visible and seems to act as background for the current viewport - it looks like the text is scrolled "over" it.

Personally I'd prefer the behavior of the cover art background and so I still think it would make sense to maybe implement the whole background image as static version of the cover art background. (i.e. div elements)
If it could be sized to fill the whole screen (like the suggested cover css method) this would look just awesome I can imagine.

Hope you can follow my explanation of the difference in appearance of the two.

**Reply by jcorporation** [reply]
Let me test it with the cover method, all but some css changes are to much work for now.

**Reply by pgzh** [reply]
Of course, it's your project and your decision! I just wanted to tell you about my experiences so you can consider changing something - and of course I wouldn't mind if you wanted to implement any of this at any later point in time.
Thanks for discussing this with me, the detailed explanations and considering changes!

**Reply by jcorporation** [reply]
I like discussing my project. Feedback is very important for me, thank you. It is so much todo and not so much time...

**Reply by jcorporation** [reply]
I changed the image background css to cover and fixed with commit 651a69f9

**Reply by pgzh** [reply]
It looks just like I imagined - thank you!
Just to let you know, it seems I cannot enable LUA in the version I built (liblua 5.3 dev package is installed - but it still says "Lua is not compiled in" in the settings page).
I also had to copy the binary manually to /usr/bin, since `build.sh pkgdebian` failed with:

`make[1]: Leaving directory '/root/myMPD/package/build/release'`
`mkdir -p debian/tmp/DEBIAN`
`cp debian/postinst debian/tmp/DEBIAN`
`cp debian/postrm debian/tmp/DEBIAN`
`cp debian/conffiles debian/tmp/DEBIAN`
`cp: cannot stat 'debian/conffiles': No such file or directory`
`make: *** [debian/rules:24: binary-arch] Error 1`
`dpkg-buildpackage: error: fakeroot debian/rules binary subprocess returned exit status 2`

I'm probably missing a package, but I am not really sure which...

Looking forward to the next stable version with this changes, thanks again :-)

**Reply by jcorporation** [reply]
The Debian packaging process is broken in the v8.0.0 branch. This will be fixed before releasing it.
Fixed with ea6f8d0c0fe813745f167e812b106f33b645abfe

**Reply by jcorporation** [reply]
> Just to let you know, it seems I cannot enable LUA in the version I built (liblua 5.3 dev package is installed - but it still says "Lua is not compiled in" in the settings page).

Please open a bug report for this issue with output of ./build.sh release, debian version, etc.
